File: C++/264. Ugly Number II.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        unordered_map<long long, int>maple;
        priority_queue<long long,vector<long long>,greater<long long> > q;
        q.push(1);
        int cnt = 0,ans = 1;
        while(cnt < n)
        {
            long long now = q.top();
            q.pop();
            if(maple.find(now)!=maple.end())
                continue;
            maple[now] = 1;
            cnt += 1;
            if(cnt == n)
                ans = now;
            q.push(now*2);
            q.push(now*3);
            q.push(now*5);
        }
        return ans;
    }
};

File: Python/264. Ugly Number II.py
class Solution:
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        ans = [1]
        i2, i3, i5 = 0, 0, 0
        for i in range(1,n):
            ans.append(min(ans[i2]*2, ans[i3]*3, ans[i5]*5));
            if ans[i] == 2*ans[i2]:
                i2 += 1
            if ans[i] == 3*ans[i3]:
                i3 += 1
            if ans[i] == 5*ans[i5]:
                i5 += 1
        return ans[n-1]